engine/common/provider/engine.go

```diff
@@ -73,81 +123,57 @@ func New(log zerolog.Logger, metrics module.EngineMetrics, net network.Network,
-// algorithm has started.
-func (e *Engine) Ready() <-chan struct{} {
-	return e.unit.Ready()
-}
-
-// Done returns a done channel that is closed once the engine has fully stopped.
-// For the consensus engine, we wait for hotstuff to finish.
-func (e *Engine) Done() <-chan struct{} {
-	return e.unit.Done()
-}
-
-// SubmitLocal submits an message originating on the local node.
-func (e *Engine) SubmitLocal(message interface{}) {
-	e.unit.Launch(func() {
-		err := e.process(e.me.NodeID(), message)
-		if err != nil {
-			engine.LogError(e.log, err)
-		}
-	})
-}
+	cm := component.NewComponentManagerBuilder()
+	cm.AddWorker(e.processQueuedRequestsShovellerWorker)
+	for i := uint(0); i < requestWorkers; i++ {
+		cm.AddWorker(e.processEntityRequestWorker)
+	}
 
-// Submit submits the given message from the node with the given origin ID
-// for processing in a non-blocking manner. It returns instantly and logs
-// a potential processing error internally when done.
-func (e *Engine) Submit(channel channels.Channel, originID flow.Identifier, message interface{}) {
-	e.unit.Launch(func() {
-		err := e.Process(channel, originID, message)
-		if err != nil {
-			engine.LogError(e.log, err)
-		}
-	})
-}
+	e.cm = cm.Build()
+	e.Component = e.cm
 
-// ProcessLocal processes an message originating on the local node.
-func (e *Engine) ProcessLocal(message interface{}) error {
-	return e.unit.Do(func() error {
-		return e.process(e.me.NodeID(), message)
-	})
+	return e, nil
 }
 
 // Process processes the given message from the node with the given origin ID in
 // a blocking manner. It returns the potential processing error when done.
-func (e *Engine) Process(channel channels.Channel, originID flow.Identifier, message interface{}) error {
-	return e.unit.Do(func() error {
-		return e.process(originID, message)
-	})
-}
-
-// process processes events for the propagation engine on the consensus node.
-func (e *Engine) process(originID flow.Identifier, message interface{}) error {
+func (e *Engine) Process(channel channels.Channel, originID flow.Identifier, event interface{}) error {
+	select {
+	case <-e.cm.ShutdownSignal():
+		e.log.Warn().
+			Hex("origin_id", logging.ID(originID)).
+			Msgf("received message after shutdown")
+		return nil
+	default:
+	}
 
 	e.metrics.MessageReceived(e.channel.String(), metrics.MessageEntityRequest)
-	defer e.metrics.MessageHandled(e.channel.String(), metrics.MessageEntityRequest)
 
-	e.unit.Lock()
-	defer e.unit.Unlock()
-
-	switch msg := message.(type) {
-	case *messages.EntityRequest:
-		return e.onEntityRequest(originID, msg)
-	default:
-		return engine.NewInvalidInputErrorf("invalid message type (%T)", message)
+	err := e.requestHandler.Process(originID, event)
+	if err != nil {
+		if engine.IsIncompatibleInputTypeError(err) {
+			e.log.Warn().
+				Hex("origin_id", logging.ID(originID)).
+				Str("channel", channel.String()).
+				Str("event", fmt.Sprintf("%+v", event)).
+				Msgf("received unsupported message type")
```

Comment:
This would mean a node sent a different message type on this channel, and none of the authorization checks detected it.  please add `Bool(logging.KeySuspicious, true).` here to make this easier to identify in the logs.

Reply:
Added flag in here: https://github.com/onflow/flow-go/pull/3404/commits/dea665dc291bb96148bc2af179fe00bb221169b5
However, an unsupported message type will be flagged and detected by the slashing consumer earlier down the pipeline.

Reply:
yea, agreed this should never happen. It would mean that somehow a valid message was passed to the wrong engine, or there was a bug somewhere